### PR TITLE
Set up publishing for Maven Central

### PR DESCRIPTION
**Pull request by drewhannay**
Initial work to publish to Maven Central

### CODE REVIEW

File: test-butler-app-core/build.gradle [hunk 1]
```diff
@@ -27,33 +23,10 @@ android {
...
-    classifier = 'sources'
-    from android.sourceSets.main.java.srcDirs
-}
-
-task javadoc(type: Javadoc) {
-    source = android.sourceSets.main.java.srcDirs
-    android.libraryVariants.matching { variant -> variant.name == 'debug'}.all { variant ->
-        classpath += project.files(android.getBootClasspath().join(File.pathSeparator))
-        classpath += variant.getCompileClasspath(null)
+        disable 'OldTargetApi', 'GradleDependency', 'PrivateApi', 'WakelockTimeout', 'DiscouragedPrivateApi'
```
Comment (chao2zhang) [inline test-butler-app-core/build.gradle:26]: Are these new lint checks introduced by AGP bump? Just to confirm

**Review by chao2zhang — APPROVED** [review]

File: test-butler-app-core/build.gradle (hunk 1 above)
Comment (drewhannay) [reply test-butler-app-core/build.gradle:26]: Yep, that's correct.

**Review by cckroets — APPROVED** [review]

**Review by chao2zhang — APPROVED** [review]

File: gradle/publishing.gradle [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+}
+
+task javadocJar(type: Jar, dependsOn: javadoc) {
+    classifier 'javadoc'
+    from javadoc.destinationDir
+}
+
+// AGP creates the components in afterEvaluate, so we need to use it too
```
Comment (DPUkyle) [inline gradle/publishing.gradle:35]: That's unfortunate but not entirely surprising; most configuration of publications must be done in an `afterEvaluate` block anyway when creation and configuration happen at different times.

File: gradle/publishing.gradle [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                        packaging = 'apk'
+                    }
+                }
+            }
+            addRepositories(publishing)
+        }
+    }
+}
+
+def addArtifacts(publication) {
```
Comment (DPUkyle) [inline gradle/publishing.gradle:96]: Nit: I suggest strongly-typing the argument as `MavenPublication`

**Review by DPUkyle — APPROVED** [review]
One thought - is Gradle Module Metadata consumable by AGP? I didn't see any mention of explicitly disabling the tasks which generate the `.module` files. We can have a separate sidebar about whether this adds value for android development or not.

File: gradle/publishing.gradle (hunk 2 above)
Comment (chao2zhang) [reply gradle/publishing.gradle:35]: There is an interesting case that we found we need nested `afterEvaluate` 😭  https://github.com/detekt/detekt/pull/3453#discussion_r573472743

**Comment by TWiStErRob** [comment]
FYI https://github.com/linkedin/test-butler/issues/113

**Pull request merged**